include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -181,18 +174,56 @@ class pool_memory_resource final
    */
   block_type expand_pool(size_t size, free_list& blocks, cudaStream_t stream)
   {
-    auto grow_size = size_to_grow(size);
-    if (grow_size == 0) {
-      RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE max pool size exceeded]",
-                    reinterpret_cast<void*>(stream),
-                    size);
-      RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
+    // Strategy. If maximum_pool_size_ is set, then grow geometrically, e.g. by halfway to the limit
+    // each time. If it is not set, grow exponentially, e.g. by doubling the pool size each time.
+    // Upon failure, attempt to back off exponentially, e.g. by half the attempted size, until
+    // either success or the attempt is less than the requested size.
+    auto try_size = size_to_grow(size);
+    while (try_size >= size) {
+      try {
+        auto b = block_from_upstream(try_size, stream);
+        current_pool_size_ += try_size;
+        return b;
+      } catch (std::exception& e) {
+        if (try_size == size) break;  // give up after we try size once
+        try_size = std::max(size, try_size / 2);
+      }
     }
-
-    current_pool_size_ += grow_size;
-    return block_from_upstream(grow_size, stream);
+    RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE max pool size exceeded]",
+                  reinterpret_cast<void*>(stream),
+                  size);
+    RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
   }
 
+  /**
+   * @brief Given a minimum size, computes an appropriate size to grow the pool.
+   *
+   * Strategy is to try to grow the pool by half the difference between the configured maximum pool
+   * size and the current pool size, if the maximum pool size is set. If it is not set, try to
+   * double the current pool size.
+   *
+   * Returns 0 if the requested size cannot be satisfied.
+   *
+   * @param size The size of the minimum allocation immediately needed
+   * @return size_t The computed size to grow the pool.
+   */
+  size_t size_to_grow(size_t size) const
+  {
+    if (maximum_pool_size_.has_value()) {
+      auto const remaining =
+        rmm::detail::align_up(maximum_pool_size_.value() - pool_size(), allocation_alignment);
+      auto const aligned_size = rmm::detail::align_up(size, allocation_alignment);
+      if (aligned_size <= remaining / 2) {
+        return remaining / 2;
+      } else if (aligned_size <= remaining) {
+        return aligned_size;
+      } else {
+        return 0;
+      }
+    } else
+      return std::max(size, pool_size());
+  };
```

Comment:
If i didn't mess up my logic, this can be simplified to the following:
```suggestion
  size_t size_to_grow(size_t size) const
  {
    if (maximum_pool_size_.has_value()) {
      auto const pool_size_delta = maximum_pool_size_.value() - pool_size();
      auto const remaining       = rmm::detail::align_up(pool_size_delta, allocation_alignment);
      auto const aligned_size    = rmm::detail::align_up(size, allocation_alignment);
      return aligned_size <= remaining ? std::max(remaining / 2, aligned_size) : 0;
    } else
      return std::max(size, pool_size());
  };
```

Reply:
> The fact that it is wrapping functions with side-effects (aka throws) makes it pretty tricky.

fwiw, Mark mentioned (and I'm extrapolating) that it's possible (though not pretty) to get around this by wrapping said function call in another function that performs the try/catch and translates the error to another form, such as a `pair<error, result>`.

Reply:
And not that I want to recommend implementing are own `expected`-like type, but the `pair<error, result>` pattern is similar to:
- Rusts `Result` type: https://doc.rust-lang.org/std/result/
- Sy Brand's `tl::expected`: https://github.com/TartanLlama/expected

and I would say having a similar type would be much clearer than having raw `std::pair`s for this use case. 

This is obviously out of the scope of this PR - but just thought I would _throw_ it out there. :joy: And note `tl::expected` is based on the proposed [`std::expected`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0323r3.pdf). And note note, `thrust::optional` is based on Sy Brand's `tl::optional`

Reply:
I don't remember saying anything about this. C++'s built-in exceptions sounds much better to me -- it is consistent with everything else in RMM.

Reply:
@codereport I also think you messed up the logic. :)  It returns remaining / 2 if size less than remaining, while the original returned remaining / 2 if size less than remaining / 2.

Reply:
> @codereport I also think you messed up the logic. :) It returns remaining / 2 if size less than remaining, while the original returned remaining / 2 if size less than remaining / 2.

I agree with your above statement - i think my thought is that they are equivalent. My visualization is the following:
```
r  = remaining
as = aligned size

< -------- r/2 ---------- r ---- >
            |             |
as =  r/2   |     as      |
```
If you look above, basically `remaining / 2` is a floor - which is what `std::max(remaining/2, aligned_size)`. My visualization might be wrong though. What do you think?

Reply:
Yeah, I think you are right. That is nicer.